pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,310 @@
+		name:   config.Bucket,
+		config: config,
+		bucket: bk,
+	}
+	return bkt, nil
+}
+
+// Iter calls f for each entry in the given directory (not recursive.). The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, objstore.DirDelim) + objstore.DirDelim
+	}
+
+	marker := alioss.Marker("")
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		objects, err2 := b.bucket.ListObjects(alioss.Prefix(dir), alioss.Delimiter(objstore.DirDelim), marker)
+		if err2 != nil {
+			return errors.Wrap(err2, "oss bucket list")
+		}
+		for _, object := range objects.Objects {
+
+			if object.Key == "" {
+				continue
+			}
+			if err := f(object.Key); err != nil {
+				return err
+			}
+		}
+		for _, object := range objects.CommonPrefixes {
+			if err := f(object); err != nil {
+				return err
+			}
+		}
+		marker = alioss.Marker(objects.NextMarker)
+		if !objects.IsTruncated {
+			break
+		}
+	}
+
+	return nil
+}
+
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+// validate checks to see the config options are set.
+func validate(conf Config) error {
+	if conf.EndPoint == "" {
+		return errors.New("no oss endpoint in config file")
+	}
+	if conf.Bucket == "" {
+		return errors.New("no oss bucket in config file")
+	}
+
+	if conf.SecretId == "" && conf.SecretKey != "" {
+		return errors.New("no oss acccess_key specified while secret_key is present in config file; both of them should be present in config.")
+	}
+
+	if conf.SecretId != "" && conf.SecretKey == "" {
+		return errors.New("no oss secret_key specified while access_key is present in config file; both of them should be present in config.")
+	}
+	return nil
+}
+
+func NewTestBucketFromConfig(t testing.TB, location string, c Config, reuseBucket bool) (objstore.Bucket, func(), error) {
+	bc, err := yaml.Marshal(c)
+	if err != nil {
+		return nil, nil, err
+	}
+	b, err := NewBucket(log.NewNopLogger(), bc, "thanos-e2e-test")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	bktToCreate := c.Bucket
+	if c.Bucket != "" && reuseBucket {
+		if err := b.Iter(context.Background(), "", func(f string) error {
+			return errors.Errorf("bucket %s is not empty", c.Bucket)
+		}); err != nil {
+			return nil, nil, errors.Wrapf(err, "oss check bucket %s", c.Bucket)
+		}
+
+		t.Log("WARNING. Reusing", c.Bucket, "Aliyun OSS bucket for OSS tests. Manual cleanup afterwards is required")
+		return b, func() {}, nil
+	}
+
+	if c.Bucket == "" {
+		src := rand.NewSource(time.Now().UnixNano())
+
+		bktToCreate = strings.Replace(fmt.Sprintf("test_%s_%x", strings.ToLower(t.Name()), src.Int63()), "_", "-", -1)
+		if len(bktToCreate) >= 63 {
+			bktToCreate = bktToCreate[:63]
+		}
+	}
+
+	if err := b.client.CreateBucket(bktToCreate, alioss.ACL(alioss.ACLPrivate), alioss.StorageClass(alioss.StorageArchive)); err != nil {
```

Comment:
Why we create archive bucket to test?

Reply:
for bucket creating test only , it is much better than standard & IA storageclass, the cold-historical data access infrequently , of cause , it is much cheaper .